Add support for verifying .debug_names in split DWARF for CUs and TUs.

clayborg opened the pull request:
This patch fixes .debug_names verification for split DWARF with no type units. It will print out an error for any name entries where we can't locate the .dwo file. It finds the non skeleton unit and correctly figures out the DIE offset in the .dwo file. If the non skeleton unit is found and yet the skeleton unit has a DWO ID, an error will be emitted showing we couldn't access the non-skeleton compile unit.

clayborg committed:
Add support for verifying .debug_names in split DWARF for CUs and TUs.

This patch fixes .debug_names verification for split DWARF with no type units. It will print out an error for any name entries where we can't locate the .dwo file. It finds the non skeleton unit and correctly figures out the DIE offset in the .dwo file. If the non skeleton unit is found and yet the skeleton unit has a DWO ID, an error will be emitted showing we couldn't access the non-skeleton compile unit.

clayborg commented:
I don't have tests yet as I am looking for the right way to create them. If anyone has any input one what I should do, please let me know. I was thinking of having tests that do:
- yaml executable + yaml .dwo file with the compilation using `-ffile-compilation-dir=.` so that llvm can find the .dwo file without worrying about paths
- yaml executable + yaml .dwp file with

Since we can't use clang and or lld  in llvm tests, I am basically looking for the best way to add test for this. If anyone can provide guidance on what the best way to do this is, I am all ears.

dwblaikie commented:
You can check the existing dwo/dwp test coverage (one easy way to find test coverage, is to break the code - like just cause the dwp or dwo lookup code to consistently fail to find the dwo/dwp file) - these days I think we go with yaml or assembly for most tests. 
For DWP files that's a bit trickier, since there's no way to generate DWP assembly files - hmm, do you need DWP testing, given this feature (split DWARF .debug_names) is orthogonal to DWP files, maybe it's sufficient to test it with DWO files? (or are there DWP-specific code paths to be tested?)

To test for just DWO files, you could checkin assembly or yaml files, maybe using the new https://llvm.org/docs/TestingGuide.html#elaborated-tests functionality to make it easier to regenerate as needed.

labath commented:
When I've needed to do that, I just created wrote the dwp section in assembly, the format is pretty simple, so you could probably just take an existing asm output from clang and slap a debug_[ct]u_index section at the back. You can use an existing DWP test as a template (e.g. test/DebugInfo/X86/dwp-v5-tu.s) 

Generally I'd recommend assembly instead of yaml as test sources, because they're easier to read in a code review (i.e., without running the input through `yaml2obj|dwarfdump`)

clayborg committed:
Add tests.

This adds a test and fixes another test to check for DIE offsets being out of range for a compile unit as this was triggered by some of the changes made for the non skeleton compile unit.

clayborg committed:
Add missing test files.

clayborg commented:
This is now fully tested. Before this patch errors would be emitted when verifying .debug_names in split DWARF because the DIE offsets wouldn't match up. This patch makes sure that works and only tests .dwo files as .dwp is just another way to access .dwo files. Let me know if there is anything missing. I have more patches needed to allow us to enable .debug_names with split DWARF, but we can't until we can verify the data is good.

ayermolo commented on llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:
```diff
@@ -1626,8 +1626,56 @@ unsigned DWARFVerifier::verifyNameIndexEntries(
+    // unit DIE of the DWARFUnit in "DU". So we need to check if the DWARFUnit
+    // has a .dwo file, but we couldn't load it.
+    DWARFDie NonSkeletonUnitDie = DU->getNonSkeletonUnitDIE();
+    if (DU->getDWOId() && DU->getUnitDIE() == NonSkeletonUnitDie) {
```
would this be a more clear check?
DWARFUnit *DWOCU =
          DwarfUnit->getNonSkeletonUnitDIE(false, AbsolutePath).getDwarfUnit();
      if (!DWOCU->isDWOUnit()) {

clayborg replied:
If the .dwo file isn't found, `DU->getNonSkeletonUnitDIE()` returns the skeleton compile unit, and then NonSkeletonUnitDie would say that it isn't a DWO file. We need to ask the skeleton unit if it has a DWO ID, and if it does _and_ the DIEs are the same, then we know we didn't find the .dwo file (in a .dwo or .dwp). So for a normal `DW_TAG_compile_unit` the `DU->getNonSkeletonUnitDIE()` will return the same DIE and DU won't have a DWO ID.

dwblaikie replied:
making `getNonSkeletonUnitDIE` return a null DWARFDIE if this is a skeleton unit but the split full unit can't be found seems reasonable - I doubt the behavior was intentional/explicitly desired, just falls out by accident

clayborg replied:
I believe it is intentional and expected behavior because as it if want all DIEs from a skeleton unit or a normal CU or TU, you call this API to try to get the DIEs regardless of if it is a skeleton unit or not. 
```
  DWARFDie getNonSkeletonUnitDIE(bool ExtractUnitDIEOnly = true,
                                 StringRef DWOAlternativeLocation = {}) {
    parseDWO(DWOAlternativeLocation);
    return DWO ? DWO->getUnitDIE(ExtractUnitDIEOnly)
               : getUnitDIE(ExtractUnitDIEOnly);
  }
```

ayermolo replied:
It is kind of weird when one calls getNonSkeletonUnitDIE, and gets SkeletonUnitDIE.
Poking around seems like only place this logic is used is in LVDWARFReader::createScopes(), and maybe in dwarfdump::collectStatsForObjectFile.

clayborg replied:
Someone can feel free to fix this quickly in a separate PR. It is used in 7 different files.
<img width="695" alt="Screenshot 2024-08-08 at 3 57 58 PM" src="https://github.com/user-attachments/assets/edba37e6-2433-427c-a6da-1698d0ad7a78">

dwblaikie replied:
> I believe it is intentional and expected behavior because as it if want all DIEs from a skeleton unit or a normal CU or TU, you call this API to try to get the DIEs regardless of if it is a skeleton unit or not.

The behavior of returning the non-split unit DIE makes sense given the name, but the behavior of returning the skeleton unit DIE if the split unit can't be found seems surprising/unintentional.

Returning the non-split unit DIE allows the "calling this function regardless of if it is a skeleton unit or not" work out, which seems good - but returning the skeleton unit DIE when the split unit isn't found seems like it'd be bad.

clayborg replied:
I agree. So you want me to fix this as part of this commit? Or in a separate commit?

ayermolo replied:
Given that it's a current behavior, maybe it's best to discuss it in follow up PR, and not let it block this one? :)

dwblaikie replied:
Sure, leave a FIXME at this call site and on the function itself.

clayborg replied:
I added the FIXME

clayborg commented:
Anything else needed for this PR?

clayborg committed:
Add a comment for a FIXME for a follow up PR.

clayborg commented:
Ok, I added the FIXME. Anything else needed for this patch?

clayborg commented:
Does anyone have time to help move this along? I have more patches coming to finish up the verification of .debug_names.

Michael137 commented:
Sorry for the late ping

Looks like the new verification error triggers in some of the tests on the DWARFv5 x86 macOS bots: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/lldb-cmake-matrix/628/execution/node/79/log/
```
********************
Unresolved Tests (3):
  lldb-api :: lang/cpp/modules-import/TestCXXModulesImport.py
  lldb-api :: lang/objc/modules-auto-import/TestModulesAutoImport.py
  lldb-api :: lang/objc/modules-objc-property/TestModulesObjCProperty.py
```

Seeing a bunch of these in the log:
```
error: Name Index @ 0x0: Entry @ 0xc93e unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc945 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc94c unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc953 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc95a unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc961 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc968 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc96f unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc976 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc97d unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc984 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc98b unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc992 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc999 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: Name Index @ 0x0: Entry @ 0xc9a0 unable to load .dwo file "None" for DWARF unit @ 0x43.
error: output verification failed for x86_64
make: *** [a.out.dSYM] Error 1
```

Trying to repro this locally now.

This is what one of the clang commands looks like:
```
"/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/bin/clang"  -std=c++11 -gdwarf-5 -isysroot "/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk" -arch x86_64  -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make/../../../../..//include -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/tools/lldb/include -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/modules-import -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make -include /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make/test_common.h  -fno-limit-debug-info    -nostdlib++ -nostdinc++ -cxx-isystem /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/include/c++/v1 -fmodules -gmodules -fmodules-cache-path=/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/lldb-test-build.noindex/module-cache-clang/lldb-api -gmodules -fcxx-modules -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/lldb-test-build.noindex/lang/cpp/modules-import/TestCXXModulesImport.test_expr_failing_import_dsym/include --driver-mode=g++ -MT main.o -MD -MP -MF main.d -c -o main.o /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/modules-import/main.cpp

"/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/bin/clang"  main.o -gdwarf-5 -isysroot "/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk" -arch x86_64  -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make/../../../../..//include -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/tools/lldb/include -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/modules-import -I/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make -include /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make/test_common.h  -fno-limit-debug-info     -L/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/lib -Wl,-rpath,/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/lib -lc++ --driver-mode=g++ -o "a.out"

ld: warning: ignoring duplicate libraries: '-lc++'

codesign --entitlements /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/packages/Python/lldbsuite/test/make/entitlements-macos.plist -s - "a.out"

"/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/lldb-build/./bin/dsymutil"  -o "a.out.dSYM" "a.out"
Verifying .debug_abbrev...
Verifying .debug_info Unit Header Chain...
Verifying .debug_types Unit Header Chain...
Verifying non-dwo Units...
Verifying unit: 1 / 2, "Foo"
Verifying unit: 2 / 2, "/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-matrix/llvm-project/lldb/test/API/lang/cpp/modules-import/main.cpp"
Verifying dwo Units...
Verifying .debug_line...
Verifying .debug_str_offsets...
Verifying .debug_names...
error: Name Index @ 0x0: Entry @ 0x9a unable to load .dwo file "None" for DWARF unit @ 0x0.
error: output verification failed for x86_64
make: *** [a.out.dSYM] Error 1
```

Note how we're compiling with `-gdwarf-5` *and* `-gmodules` in all of the failing tests. So probably `-gmodules` related

Michael137 commented:
So looks like the offending DWARF is this:
```
.debug_info contents:
0x00000000: Compile Unit: length = 0x00000033, format = DWARF32, version = 0x0005, unit_type = DW_UT_compile, abbr_offset = 0x0000, addr_size = 0x08 (next unit at 0x00000037)

0x0000000c: DW_TAG_compile_unit
              DW_AT_producer    ("clang version 20.0.0git (https://github.com/Michael137/llvm-project.git fb326019769e448f9effe3c8b10e6df8e7e9ce71)")
              DW_AT_language    (DW_LANG_C_plus_plus_11)
              DW_AT_name        ("Foo")
              DW_AT_LLVM_sysroot        ("/path/to/sdk")
              DW_AT_APPLE_sdk   ("some.sdk")
              DW_AT_str_offsets_base    (0x00000008)
              DW_AT_stmt_list   (0x00000000)
              DW_AT_comp_dir    ("/Users/michaelbuch/Git/llvm-worktrees/playground/lldb-test-build.noindex/lang/cpp/modules-import/TestCXXModulesImport.test_expr_dsym")
              DW_AT_GNU_dwo_id  (0xb0d9e392fcae5b5c)
...

0x00000037: Compile Unit: length = 0x0000006c, format = DWARF32, version = 0x0005, unit_type = DW_UT_compile, abbr_offset = 0x0000, addr_size = 0x08 (next unit at 0x000000a7)

0x00000043: DW_TAG_compile_unit
              DW_AT_producer    ("clang version 20.0.0git (https://github.com/Michael137/llvm-project.git fb326019769e448f9effe3c8b10e6df8e7e9ce71)")
              DW_AT_language    (DW_LANG_C_plus_plus_11)
              DW_AT_name        ("/Users/michaelbuch/Git/llvm-worktrees/playground/lldb/test/API/lang/cpp/modules-import/main.cpp")
              DW_AT_LLVM_sysroot        ("/path/to/sdk")
              DW_AT_APPLE_sdk   ("some.sdk")
              DW_AT_str_offsets_base    (0x00000008)
              DW_AT_stmt_list   (0x0000003d)
              DW_AT_comp_dir    ("/Users/michaelbuch/Git/llvm-worktrees/playground/lldb-test-build.noindex/lang/cpp/modules-import/TestCXXModulesImport.test_expr_dsym")
              DW_AT_low_pc      (0x0000000100003f88)
              DW_AT_high_pc     (0x0000000100003fa8)
              DW_AT_addr_base   (0x00000008)
```

Note that the first CU has a `DW_AT_GNU_dwo_id` but there's no associated "DWO file" (which in the `gmodules` case is a CU with a matching `DW_AT_GNU_dwo_id` and a path to the `.pcm` in `DW_AT_dwo_name`).

AFAIU, this happens when we compile with `-gmodules`, followed by `dsymutil` (which, mostly (?), undoes `-gmodules`). But we're left with a `DW_AT_GNU_dwo_id` in the main CU. So with the new verifier check added in this PR:
```
if (DU->getDWOId() && DU->getUnitDIE() == NonSkeletonUnitDie) {
```
, we do have a `getDWOId`, but `DU->DWO` is empty, because there is none. And we claim to have failed to load the `.dwo` file.

Any idea on the best way forward here? @adrian-prantl @clayborg 

Maybe `dsymutil` shouldn't be cloning over the `DW_AT_GNU_dwo_id`? Locally that fixed things for me.

This new verifier check is causing a lot of our modules builds to fail.

Michael137 commented:
Proposed fix in: https://github.com/llvm/llvm-project/pull/105186
